player/lua/osc.lua

```diff
@@ -2609,6 +2612,7 @@ mp.observe_property("fullscreen", "bool",
     function(name, val)
         state.fullscreen = val
         request_init_resize()
+        update_margins()
```

Comment:
The order is a bit confusing. `update_margins()` is called last, but in fact it's performed first, because `request_init_resize()` results in an a-synchronous `tick()` (even if immediate - always with timeout 0).

Maybe this should use a `state.marginsREQ` (similar to `state.initREQ`) before `request_init_resize()` to keep a logical order, and applied at the begining of `tick()` - before `state.initREQ` is applied (which is at `render()`), like so maybe?
```lua
    function(name, val)
        state.fullscreen = val
        state.marginsREQ = true
        request_init_resize()
    end
```

Reply:
Or maybe make `request_init_resize()` take an argument of whether to also init margins, which sets `state.marginsREQ`. The other calls to request_init_resize wouldn't need modification because missing argument is falsy.

Reply:
You know I just realized the fix is as simple as calling `reset_margins()` at `render_wipe()`

The problem we're trying to fix is that the margins stay when the OSC is removed. So remove the margins when doing so. It's funny how the fix is this simple and logical and I didn't realize it earlier.

Since we replaced the empty string clear with `render_wipe()` at `tick()` like you recommended for the other fix it makes this easy

I tested it and it works for both show* opts when using boxvideo:
https://github.com/ossifrage/mpv/commit/19c736ea77e6dee2dfe10e63bff872ae76ae85be

What do you think?

Reply:
> What do you think?

Interesting, and I wouldn't mind a fix based on this, but not at this form, because `render_wipe` is currently very light and technically (at lower levels) is a no-op if the the overlay is not visible. Theoretically we could even call wipe on every tick before render (or even whenever `state.enabled` is false).

But your approach does real work on every wipe beyond removing the overlay.

Reply:
Also, logically, I think margins is not related to wipe. It's only related to wipe in the context of resetting the margins (which is what this suggested fix does, but `update_margins` does more than just reset).

Reply:
Another problem I see with it is if the user is using custom margins, any `render_wipe()` call will reset them even if they're not using boxvideo.

What about calling `update_margins()` at `render_wipe` instead and leaving the show* opt conditions in `update_margins()` like they're currently in this PR?

Reply:
> Or maybe make request_init_resize() take an argument of whether to also init margins

I do understand that this creates slight redundancy, and maybe a better apprroach would be to move the `update_margins` decision-making elsewhere, so that it's decided only at one place.

We can do that now or later or never or at the states rework. Or just do something along the lines of my initial suggestions for now. Take a pick.

Reply:
How does this look: https://github.com/ossifrage/mpv/commit/770bda73118d5772153e78c4837c16b9a52b21fc

Reply:
> How does this look: [ossifrage@770bda7](https://github.com/ossifrage/mpv/commit/770bda73118d5772153e78c4837c16b9a52b21fc)

Looks good to me. We can take it as is if you're happy with it.

Reply:
> Looks good to me. We can take it as is if you're happy with it.

Actually, `state.marginsREQ` is serviced after `state.enabled` is tested. I presume there are no cases where it's disabled but we still want to update the margins, right? If there are, or as a safety measre, we could swap these. Up to you.

Reply:
> I presume there are no cases where it's disabled but we still want to update the margins, right?

Not currently, no. At least not that I can see. Toggling the visibility works since it calls `update_margins()` anyway and disabling the OSC calls `reset_margins` at `shutdown`

> as a safety measre, we could swap these

Might be a good idea anyway especially if there's subsequent work on the states. I can imagine a scenario where `state.osc_visible` is used instead to determine the margins for the show* opts and if `state.enabled` is set to false before `tick()` then the margins won't reset. It doesn't hurt to swap them

Reply:
> I can imagine a scenario where `state.osc_visible` is used instead to determine

Yeah, though a rework of the states will almost certainly also rework all the big yes/no decisions-making into one place instead of keeping it spread all over the place. So it's unlikely to get missed out during such rework.

Anyway, I also think it's safer, but up to you.

Reply:
Okay, I've updated the PR